src/index.ts

```diff
@@ -101,147 +60,174 @@ function computeCost(job: Job, providerPrice: ProviderPricesStruct): bigint {
-    jsonRpcProvider: JsonRpcProvider = new JsonRpcProvider(
-      "https://testnet.deepsquare.run/rpc",
-      {
-        name: "DeepSquare Testnet",
-        chainId: 179188,
-      }
-    ),
+    publicClient: PublicClient = createPublicClient({
+      transport: http("https://testnet.deepsquare.run/rpc"),
+      chain: deepSquareChain,
+    }),
     loggerClientFactory: () => ILoggerAPIClient = createLoggerClient
-  ): Promise<DeepSquareClient> {
-    // Use a authenticated client if there is a key, else don't.
-    const signerOrProvider = privateKey
-      ? new Wallet(privateKey, jsonRpcProvider)
-      : jsonRpcProvider;
-    const metaScheduler = MetaScheduler__factory.connect(
-      metaschedulerAddr,
-      signerOrProvider
-    );
-    const creditAddr = await metaScheduler.credit();
-    const credit = IERC20__factory.connect(creditAddr, signerOrProvider);
-
-    const providerAddr = await metaScheduler.providerManager();
-    const providerManager = IProviderManager__factory.connect(
-      providerAddr,
-      signerOrProvider
-    );
-
-    return new DeepSquareClient(
-      signerOrProvider,
-      metaScheduler,
-      credit,
-      providerManager,
-      new GraphQLClient(sbatchServiceEndpoint),
-      loggerClientFactory
-    );
+  ) {
+    this.publicClient = publicClient;
+    if (privateKey) {
+      this.wallet = createWalletClient({
+        account: privateKeyToAccount(privateKey),
+        chain: deepSquareChain,
+        transport: http("https://testnet.deepsquare.run/rpc"),
+      });
+    } else {
+      this.wallet = wallet;
+    }
+    this.sbatchServiceClient = new GraphQLClient(sbatchServiceEndpoint);
+    this.loggerClientFactory = loggerClientFactory;
+    this.metaSchedulerAddr = metaschedulerAddr;
   }
 
   /**
    * This method allows the DeepSquare Grid to consume a specific amount of credits from the client's
    * account for running jobs. The credits act as the payment medium for the computational resources used.
    *
-   * @param {BigNumber} amount - The amount of credits the client approves to be used for job execution.
+   * @param {bigint} amount - The amount of credits the client approves to be used for job execution.
    *   This amount is represented as a BigNumber, which helps handle very large numbers safely in JavaScript.
    *
    * Note: Be aware that the amount is in the smallest unit of the currency, like wei for Ethereum.
    *
    * The approval is given to the MetaScheduler smart contract, which manages the job execution on the Grid.
    */
-  async setAllowance(amount: BigNumber) {
-    await this.credit.approve(
-      this.metaScheduler.address,
-      parseUnits(amount.toString(), "ether")
-    );
+  async setAllowance(amount: bigint) {
+    if (!this.wallet) {
+      throw new Error(
+        "Client has been instanced without wallet client and is therefore unable to execute write operations"
+      );
+    }
+    if (!this.creditAddr) {
```

Comment:
This looks like a lazy singleton pattern. Can you refactor this and extract this for futur proofing (because we will probably use it for other functions like `getAllowance`)? Something like this:

```ts
async getCreditAddr() Promise<Hex> {
  if (!this.creditAddr) {
    this.creditAddr = await this.publicClient.readContract({
        address: this.metaSchedulerAddr,
        abi: MetaSchedulerAbi,
        functionName: "credit",
      });
  }
  return this.creditAddr;
}
```